app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/SenderDevice.kt

```diff
@@ -0,0 +1,106 @@
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.withContext
+import org.kiwix.kiwixmobile.core.BuildConfig
+import org.kiwix.kiwixmobile.localFileTransfer.WifiDirectManager.Companion.copyToOutputStream
+import java.io.IOException
+import java.net.InetAddress
+import java.net.InetSocketAddress
+import java.net.Socket
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Once the handshake between the two connected devices has taked place, this async-task is used
+ * on the sender device to transfer the file to the receiver device at the FILE_TRANSFER_PORT port.
+ *
+ * It takes in the uri of all the files to be shared. For each file uri, creates a new connection &
+ * copies all the bytes from input stream of the file to the output stream of the receiver device.
+ * Also changes the status of the corresponding FileItem on the list of files for transfer.
+ *
+ * A single task is used by the sender for the entire transfer
+ */
+private const val TIME_OUT = 15000
+
+internal class SenderDevice(
+  activity: Activity,
+  private val wifiDirectManager: WifiDirectManager,
+  val fileReceiverDeviceAddress: InetAddress
+) {
+  private val contentResolver: ContentResolver = activity.contentResolver
+  private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
+  suspend fun send(fileItems: List<FileItem?>) = withContext(ioDispatcher) {
+    delayForSlowReceiverDevicesToSetupServer()
+    val hostAddress =
+      fileReceiverDeviceAddress.hostAddress
+    var isTransferErrorFree = true
+    fileItems.asSequence()
+      .takeWhile { isActive } // checks if coroutine is live
+      .forEachIndexed { fileIndex, fileItem ->
+        try {
+          Socket().use { socket ->
+            contentResolver.openInputStream(fileItem?.fileUri!!).use { fileInputStream ->
+              socket.bind(null)
+              socket.connect(
+                InetSocketAddress(hostAddress, WifiDirectManager.FILE_TRANSFER_PORT),
+                TIME_OUT
+              )
+              Log.d(TAG, "Sender socket connected to server - " + socket.isConnected)
+              publishProgress(fileIndex, FileItem.FileStatus.SENDING)
+              val socketOutputStream = socket.getOutputStream()
+              copyToOutputStream(fileInputStream!!, socketOutputStream)
+              if (BuildConfig.DEBUG) Log.d(TAG, "Sender: Data written")
+              publishProgress(fileIndex, FileItem.FileStatus.SENT)
+            }
+          }
+        } catch (e: IOException) {
+          Log.e(TAG, e.message)
+          e.printStackTrace()
+          isTransferErrorFree = false
+          publishProgress(fileIndex, FileItem.FileStatus.ERROR)
+        }
+      }
+
+    return@withContext isTransferErrorFree
+  }
+
+  private suspend fun publishProgress(fileIndex: Int, fileStatus: FileItem.FileStatus) {
+    withContext(Dispatchers.Main) {
+      wifiDirectManager.changeStatus(fileIndex, fileStatus)
+    }
+  }
+
+  @SuppressWarnings("MagicNumber")
```

Comment:
we can get rid of this and add another const val with `TIME_OUT`, maybe `FOR_SLOW_RECEIVERS` and then we can delete this method and this comment and now it reads just fine in the original function `delay(FOR_SLOW_RECEIVERS)`